File: BioAI/Java/BioAI.java
package com.bioai.runtime;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.IOException;
import java.util.Map;

/**
 * Java-Wrapper für die BioAI-Engine (v0.7.6).
 * Implementiert AutoCloseable für automatische Ressourcenfreigabe.
 */
public class BioBrainInstance implements AutoCloseable {

    // Definition der nativen Schnittstelle
    private interface BioAILibrary extends Library {
        // Name der Bibliothek (ohne Endung, z.B. BioAI_ULTRA)
        BioAILibrary INSTANCE = Native.load("BioAI_ULTRA", BioAILibrary.class);

        Pointer API_CreateBrain(long key);
        void API_FreeBrain(Pointer brainPtr);
        void API_SetMode(Pointer brainPtr, int mode);
        long API_Update(Pointer brainPtr, long[] inputs, int count);
        long API_Simulate(Pointer brainPtr, long[] inputs, int count, int depth);
        void API_Feedback(Pointer brainPtr, float reward, long action);
        void API_Teach(Pointer brainPtr, long input, long action, float weight);
        float API_Inspect(Pointer brainPtr, long input, long action);
        Pointer API_Serialize(Pointer brainPtr, IntByReference outSize);
        Pointer API_Deserialize(byte[] data, int size);
        void API_FreeBuffer(Pointer buffer);
    }

    private Pointer brainHandle;
    private final long licenseKey;

    /**
     * Initialisiert eine neue Brain-Instanz durch Laden des Schlüssels aus einer JSON-Datei.
     * @param jsonPath Pfad zur ISS-generierten key.json.
     */
    public BioBrainInstance(String jsonPath) throws IOException {
        // 1. Key aus JSON extrahieren
        ObjectMapper mapper = new ObjectMapper();
        Map<String, String> keyData = mapper.readValue(new File(jsonPath), Map.class);
        String rawKey = keyData.get("customer_key");

        // 2. Hex-String zu long konvertieren (Entfernt C-Suffixe)
        this.licenseKey = Long.parseUnsignedLong(rawKey.replace("0x", "").replace("ULL", ""), 16);

        // 3. Native Instanz erzeugen
        this.brainHandle = BioAILibrary.INSTANCE.API_CreateBrain(this.licenseKey);
        
        if (this.brainHandle == null) {
            throw new RuntimeException("BioAI: Initialisierung der nativen Instanz fehlgeschlagen.");
        }
    }

    /**
     * Schaltet zwischen Training (0) und Produktion (1) um.
     */
    public void setMode(int mode) {
        BioAILibrary.INSTANCE.API_SetMode(brainHandle, mode);
    }

    /**
     * Verarbeitet Eingaben und liefert die optimale Aktion in O(1).
     */
    public long update(long[] inputs) {
        return BioAILibrary.INSTANCE.API_Update(brainHandle, inputs, inputs.length);
    }

    /**
     * Führt eine Kausalitäts-Simulation durch.
     */
    public long simulate(long[] inputs, int depth) {
        return BioAILibrary.INSTANCE.API_Simulate(brainHandle, inputs, inputs.length, depth);
    }

    /**
     * Passt das Verhalten über Reinforcement Learning an.
     */
    public void feedback(float reward, long action) {
        BioAILibrary.INSTANCE.API_Feedback(brainHandle, reward, action);
    }

    /**
     * Injiziert einen festen Reflex direkt in das LTM.
     */
    public void teach(long input, long action, float weight) {
        BioAILibrary.INSTANCE.API_Teach(brainHandle, input, action, weight);
    }

    /**
     * Liest ein gelerntes Gewicht unter De-Salting aus.
     */
    public float inspect(long input, long action) {
        return BioAILibrary.INSTANCE.API_Inspect(brainHandle, input, action);
    }

    /**
     * Erzeugt einen binären Snapshot des Wissenszustands.
     */
    public byte[] serialize() {
        IntByReference sizeRef = new IntByReference();
        Pointer buffer = BioAILibrary.INSTANCE.API_Serialize(brainHandle, sizeRef);
        
        if (buffer == null) return null;

        int size = sizeRef.getValue();
        byte[] result = buffer.getByteArray(0, size);
        
        // Native Freigabe des Puffers
        BioAILibrary.INSTANCE.API_FreeBuffer(buffer);
        return result;
    }

    /**
     * Schließt die Instanz und gibt nativen Speicher frei.
     */
    @Override
    public void close() {
        if (brainHandle != null) {
            BioAILibrary.INSTANCE.API_FreeBrain(brainHandle);
            brainHandle = null;
        }
    }
}

File: BioAI/JavaScript/BioAI.js
const ffi = require('ffi-napi');
const ref = require('ref-napi');
const fs = require('fs');
const path = require('path');

// Definition der nativen Typen für JavaScript/Node.js
const uint64 = ref.types.uint64;
const voidPtr = ref.refType(ref.types.void);

class BioBrainInstance {
    /**
     * Erzeugt eine neue Instanz und lädt den Schlüssel aus der key.json.
     * @param {string} jsonPath Pfad zur ISS-generierten key.json.
     * @param {string} dllPath Pfad zur nativen DLL/SO.
     */
    constructor(jsonPath, dllPath = "BioAI_ULTRA.dll") {
        // 1. Bibliothek laden
        this._lib = ffi.Library(path.resolve(dllPath), {
            'API_CreateBrain': [voidPtr, [uint64]],
            'API_FreeBrain': ['void', [voidPtr]],
            'API_SetMode': ['void', [voidPtr, 'int']],
            'API_Update': [uint64, [voidPtr, ref.refType(uint64), 'int']],
            'API_Simulate': [uint64, [voidPtr, ref.refType(uint64), 'int', 'int']],
            'API_Feedback': ['void', [voidPtr, 'float', uint64]],
            'API_Teach': ['void', [voidPtr, uint64, uint64, 'float']],
            'API_Inspect': ['float', [voidPtr, uint64, uint64]],
            'API_Serialize': [voidPtr, [voidPtr, ref.refType('int')]],
            'API_FreeBuffer': ['void', [voidPtr]]
        });

        // 2. Key aus JSON laden
        const keyData = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
        const rawKey = keyData.customer_key.replace("ULL", ""); // Entfernt C-Suffix
        this.licenseKey = BigInt(rawKey);

        // 3. Brain instanziieren
        this._brainHandle = this._lib.API_CreateBrain(this.licenseKey);
        if (this._brainHandle.isNull()) {
            throw new Error("BioAI: Initialisierung der Brain-Instanz fehlgeschlagen.");
        }
    }

    /** 0 = Training, 1 = Produktion (Fixed Structure). */
    setMode(mode) {
        this._lib.API_SetMode(this._brainHandle, mode);
    }

    /** Verarbeitet Inputs und liefert die optimale Aktion in O(1). */
    update(inputs) {
        const inputBuffer = Buffer.alloc(inputs.length * 8);
        for (let i = 0; i < inputs.length; i++) {
            inputBuffer.writeBigUInt64LE(BigInt(inputs[i]), i * 8);
        }
        return this._lib.API_Update(this._brainHandle, inputBuffer, inputs.length);
    }

    /** Führt eine Kausalitäts-Simulation durch. */
    simulate(inputs, depth) {
        const inputBuffer = Buffer.alloc(inputs.length * 8);
        for (let i = 0; i < inputs.length; i++) {
            inputBuffer.writeBigUInt64LE(BigInt(inputs[i]), i * 8);
        }
        return this._lib.API_Simulate(this._brainHandle, inputBuffer, inputs.length, depth);
    }

    /** Passt das Verhalten über Reinforcement Learning an. */
    feedback(reward, action) {
        this._lib.API_Feedback(this._brainHandle, reward, BigInt(action));
    }

    /** Injiziert eine harte Regel (Reflex) direkt in das LTM. */
    teach(input, action, weight) {
        this._lib.API_Teach(this._brainHandle, BigInt(input), BigInt(action), weight);
    }

    /** Liest ein gelerntes Gewicht unter De-Salting aus. */
    inspect(input, action) {
        return this._lib.API_Inspect(this._brainHandle, BigInt(input), BigInt(action));
    }

    /** Erzeugt einen Snapshot des Gehirns. */
    serialize() {
        const sizePtr = ref.alloc('int');
        const buffer = this._lib.API_Serialize(this._brainHandle, sizePtr);
        if (buffer.isNull()) return null;

        const size = sizePtr.deref();
        const result = Buffer.from(ref.reinterpret(buffer, size, 0));

        // Native Freigabe
        this._lib.API_FreeBuffer(buffer);
        return result;
    }

    /** Gibt alle nativen Ressourcen frei. */
    close() {
        if (!this._brainHandle.isNull()) {
            this._lib.API_FreeBrain(this._brainHandle);
            this._brainHandle = ref.NULL_POINTER;
        }
    }
}

module.exports = BioBrainInstance;